core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -787,9 +803,37 @@ class FetchSessionCache(private val maxEntries: Int,
     }
   }
 }
+object FetchSessionCache {
+  private[server] val metricsGroup = new KafkaMetricsGroup(classOf[FetchSessionCache])
+  private val counter = new AtomicLong(0)
+}
+
+class FetchSessionCache(private val cacheShards: Seq[FetchSessionCacheShard]) {
+  // Set up metrics.
+  FetchSessionCache.metricsGroup.newGauge(FetchSession.NUM_INCREMENTAL_FETCH_SESSIONS, () => cacheShards.map(_.size).sum)
+  FetchSessionCache.metricsGroup.newGauge(FetchSession.NUM_INCREMENTAL_FETCH_PARTITIONS_CACHED, () => cacheShards.map(_.totalPartitions).sum)
+
+  def getCacheShard(sessionId: Int): FetchSessionCacheShard = {
+    val shard = sessionId / cacheShards.head.sessionIdRange
+    cacheShards(shard)
+  }
+
+  // Returns the shard in round-robin
+  def getNextCacheShard: FetchSessionCacheShard = {
+    val shardNum = (FetchSessionCache.counter.getAndIncrement() % size).toInt
```

Comment:
As `int` is enough to this case, maybe we can use `AtomicInteger`?

Reply:
I used `AtomicLong` to practically rule out an overflow but found `Utils.toPositive` which is used by `RoundRobinPartitioner` :) Updated to use an `AtomicInteger` and also added some test to ensure round-robin allocations.